packages/react-server/src/ReactFizzHooks.js

```diff
@@ -577,22 +587,76 @@ function useOptimistic<S, A>(
   return [passthrough, unsupportedSetOptimisticState];
 }
 
+function isSameFormStateInstance(k1: KeyNode, k2: KeyNode): boolean {
```

Comment:
If we accepted a string as the argument to Fizz instead, this could just be `JSON.stringify(key) === formState.key`. That would also allow us to optionally apply a hash to the key.